### PR TITLE
Generate random names for resources consumed by SUC

### PR DESCRIPTION
**Pull request by ipetrov117**
Problem:
- Currently SUC resources generated by the `upgrade-controller` have hardcoded names which can lead to collisions (or SUC simply refusing to reinitiate an upgrade) when an edit happens on the `UpgradePlan`. This is applicable for other use-cases apart from edit, for example when a failure happens and a user needs to rerun the whole plan (currently, by recreating it).

Suggested approach:
- All SUC created resources should have a random suffix appended to their names, which will result in no collision of resources during edits/recreations.

How is the suffix generated?
- The suffix is generated by generating 5 random byte values and turning them to their hexadecimal equivalent. As it is with every programmable "randomisation" there is a chance of duplicate suffix generation. The chance is **extremely low**, but in order to be safe a check is implemented that validates if the old suffix is the same as the new suffix. 

Why can't we use `metadata.GenerateName`?
- Due to the nature of how the `ugprade-controller` operates it would be hard to implement the `metadata.GenerateName` usage without having to go over unnecessary loops and checks..

Naming with suggested approach:
- `UpgradePlan`:
```yaml
apiVersion: lifecycle.suse.com/v1alpha1
  kind: UpgradePlan
  ...
  status:
    conditions:
    - lastTransitionTime: "2024-08-30T07:42:22Z"
      message: All cluster nodes are upgraded
      reason: Succeeded
      status: "True"
      type: OSUpgraded
    - lastTransitionTime: "2024-08-30T07:43:30Z"
      message: All cluster nodes are upgraded
      reason: Succeeded
      status: "True"
      type: KubernetesUpgraded
    - lastTransitionTime: "2024-08-30T07:43:30Z"
      message: Chart rancher upgrade succeeded
      reason: Succeeded
      status: "True"
      type: RancherUpgraded
    ....
    sucNameSuffix: e94edd4046
```
- SUC Plans:
```bash
mgmt-cluster-network:~/upgrade-controller # k get plans -n cattle-system
NAME                                      IMAGE                                 CHANNEL   VERSION
control-plane-sl-micro-6-0-e94edd4046     registry.suse.com/bci/bci-base:15.5             3.0.1
control-plane-v1-28-9-rke2r1-e94edd4046   rancher/rke2-upgrade                            v1.28.9+rke2r1
```
- Upgrade pods:
```bash
mgmt-cluster-network:~/upgrade-controller # k get pods -n cattle-system
NAME                                                              READY   STATUS      RESTARTS     AGE
apply-control-plane-sl-micro-6-0-e94edd4046-on-mgmt-clust-wsqtg   0/1     Completed   0            2m17s
apply-control-plane-v1-28-9-rke2r1-e94edd4046-on-mgmt-clu-chqdg   0/1     Completed   0            2m
rancher-6cd44cb4b5-8xwvp                                          1/1     Running     0            8d
rancher-6cd44cb4b5-cdbcn                                          1/1     Running     0            8d
rancher-6cd44cb4b5-gnpp8                                          1/1     Running     1 (8d ago)   8d
rancher-webhook-84df44577b-kc7g9                                  1/1     Running     0            8d
system-upgrade-controller-6f86d6d4df-qlwtq                        1/1     Running     2 (8d ago)   8d
```
- OS upgrade secret:
```bash
mgmt-cluster-network:~/upgrade-controller # k get secrets -n cattle-system
NAME                                         TYPE                                  DATA   AGE
...
os-upgrade-secret-sl-micro-6-0-e94edd4046    Opaque                                1      2m21s
...
```

### CODE REVIEW

File: internal/controller/upgradeplan_controller.go [hunk 1]
```diff
@@ -169,6 +169,19 @@ func (r *UpgradePlanReconciler) reconcileNormal(ctx context.Context, upgradePlan
...
 
 	if len(upgradePlan.Status.Conditions) == 0 {
+		suffix, err := upgrade.GenerateSuffix()
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("generating suffix: %w", err)
+		}
+
+		// validate that the generated suffix is not the same
+		// as the current suffix
+		if suffix == upgradePlan.Status.SUCNameSuffix {
```
Comment (ipetrov117) [inline internal/controller/upgradeplan_controller.go:179]: 99% of the times we won't hit this, added it just to be sure that even if we did, we will reconcile and generate a new suffix.

File: internal/controller/upgradeplan_controller.go [hunk 2]
```diff
@@ -169,6 +169,19 @@ func (r *UpgradePlanReconciler) reconcileNormal(ctx context.Context, upgradePlan
...
+			return ctrl.Result{Requeue: true}, nil
```
Comment (atanasdinov) [inline internal/controller/upgradeplan_controller.go:180]: Do you think it's better to requeue with an error instead in order not to fail _silently_ here?

**Review by atanasdinov — COMMENTED** [review]
LGTM!

File: internal/controller/upgradeplan_controller.go (hunk 2 above)
Comment (ipetrov117) [reply internal/controller/upgradeplan_controller.go:180]: I am not sure that the user needs to be aware that we hit a duplicate on an internal property.

**Review by atanasdinov — APPROVED** [review]

**Pull request merged**